File: python/src/beval/subject.py
"""Subject normalization for the beval framework.

Wraps system output into the normalized Subject interface.
See SPEC.md §7.2 (The Subject).
"""

from __future__ import annotations

from typing import Any

from beval.types import Subject


def normalize_subject(
    *,
    input: str | list[dict[str, Any]],  # noqa: A002
    output: str | list[dict[str, Any]],  # noqa: A002
    completion_time: float,
    tool_calls: list[dict[str, Any]] | None = None,
    spans: list[Any] | None = None,
    metadata: dict[str, Any] | None = None,
    stage: int | None = None,
    stage_name: str | None = None,
    prior_subject: Subject | None = None,
) -> Subject:
    """Create a normalized Subject from system output. See SPEC §2.2.

    The returned Subject is frozen (immutable) to prevent graders from
    modifying shared state.
    """
    return Subject(
        input=input,
        output=output,
        completion_time=completion_time,
        tool_calls=tool_calls or [],
        spans=spans or [],
        metadata=metadata or {},
        stage=stage,
        stage_name=stage_name,
        prior_subject=prior_subject,
    )

File: dotnet/src/Beval/Models/Subject.cs
namespace Beval.Models;

/// <summary>
/// Normalized system output exposed to graders. See SPEC §7.2.
/// </summary>
/// <param name="Query">The input query.</param>
/// <param name="Answer">The system response text.</param>
/// <param name="CompletionTime">Time in seconds to produce the answer.</param>
/// <param name="DocumentsRetrieved">Number of documents retrieved.</param>
/// <param name="CitationsCount">Number of citations in the answer.</param>
/// <param name="SourcesUsed">Sources referenced by the system.</param>
/// <param name="Spans">OpenTelemetry spans from the system execution.</param>
/// <param name="Metadata">Arbitrary metadata from the system.</param>
/// <param name="Stage">Stage index for multi-stage cases.</param>
/// <param name="StageName">Stage name for multi-stage cases.</param>
/// <param name="PriorSubject">Subject from the previous stage.</param>
public record Subject(
    string Query,
    string Answer,
    double CompletionTime,
    int DocumentsRetrieved = 0,
    int CitationsCount = 0,
    IReadOnlyList<string>? SourcesUsed = null,
    IReadOnlyList<object>? Spans = null,
    IReadOnlyDictionary<string, object>? Metadata = null,
    int? Stage = null,
    string? StageName = null,
    Subject? PriorSubject = null);
